File: examples/node/paper_trade.js
// Paper-trade differentiator demo. Run: node paper_trade.js
"use strict";

const { Exchange, OrderRequest } = require("wickra-exchange");

const ex = Exchange.paper({ USDT: 100000 }, 1, 5, 10);
console.log("venue:", ex.name());
ex.setPrice("BTC/USDT", 20000);

const order = ex.placeOrder(OrderRequest.marketBuy("BTC/USDT", 1));
console.log(`filled at ${order.averagePrice} (status ${order.status})`);

for (const [asset, free] of Object.entries(ex.balances())) {
  console.log(`  ${asset} free: ${free}`);
}

for (const event of ex.pollEvents()) {
  console.log("event:", event.kind);
}

File: examples/rust/src/paper_trade.rs
//! The differentiator demo: an offline paper account fills orders deterministically
//! through the same `Exchange` API a live venue uses.
//!
//! Run with: `cargo run -p wickra-exchange-examples --bin paper_trade`

use rust_decimal_macros::dec;
use wickra_exchange::{Exchange, Execution, MarketData, OrderRequest, PaperExchange, Symbol};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let market = Symbol::new("BTC", "USDT");

    let mut exchange = PaperExchange::new()
        .with_fees(dec!(1), dec!(5)) // maker / taker basis points
        .with_slippage_bps(dec!(10))
        .with_balance("USDT", dec!(100000));
    exchange.set_price(&market, dec!(20000));

    println!("venue: {}", exchange.name());

    let order = exchange.place_order(&OrderRequest::market_buy(market.clone(), dec!(1)))?;
    println!(
        "filled {} {} at {:?} (status {:?})",
        order.filled_quantity, market, order.average_price, order.status
    );

    for balance in exchange.balances()? {
        println!("  {} free: {}", balance.asset, balance.free);
    }

    // Execution events flow through the same pull loop as market data.
    for event in exchange.poll_events() {
        println!("event: {event:?}");
    }

    Ok(())
}
